Smart block appender

draganescu opened the pull request:
## Description
Closes #16659. 

## How has this been tested?
- checkout the branch 
- enable the navigation block in Gutenberg > Experiments
- insert a Navigation block
- try to insert with the appender

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

draganescu committed:
if thre is only one there is only one

mtias commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -83,7 +93,23 @@ class Inserter extends Component {
+						className={ classnames( 'block-editor-button-block-appender' ) }
+						onClick={ createIfOne }
+					>
+						<span className="screen-reader-text">{ __( 'Add Block' ) }</span>
```
We should try to use the name of the child block here.

draganescu replied:
Good idea, will try

draganescu replied:
Ok, there is a baked in way, but I am not too proud of it considering a split on a slash and a replace of hyphens is necessary.

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -102,10 +128,62 @@ class Inserter extends Component {
+			} = select( 'core/block-editor' );
+			const parentBlockListSettings = getBlockListSettings( rootClientId );
+			const parentAllowedBlocks = get( parentBlockListSettings, [ 'allowedBlocks' ] );
+			if ( parentAllowedBlocks.length > 1 ) {
```
Is there ever a chance that `allowedBlocks` is missing? Could instead do:

```js
const isOnlyOneAllowedBlock = get( parentBlockListSettings, [ 'allowedBlocks', 'length' ], 0 ) === 1
```

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -102,10 +128,62 @@ class Inserter extends Component {
+	} ),
+	withDispatch( ( dispatch, ownProps, { select } ) => {
+		// eslint-disable-next-line no-restricted-syntax
+		function getInsertionIndex() {
```
The original `getInsertionIndex`, in `InserterMenu`, should be made into a selector — unless there's a reason not to.

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1328,6 +1328,29 @@ export const hasInserterItems = createSelector(
+ *
+ * @return {boolean} True if there is one item available, false if zero or more than one.
+ */
+export const hasOneAllowedItem = createSelector(
```
Was there a reason to cache this selector? It neither returns a new reference upon each new call, nor is it expensive (its dependency, `getBlockListSettings`, is a simple object access).

draganescu replied:
no there was no reason other than me not having enough experience with wp data and not paying attention to what createselector does :)

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1328,6 +1328,29 @@ export const hasInserterItems = createSelector(
+		if ( rootClientId ) {
+			const parentBlockListSettings = getBlockListSettings( state, rootClientId );
+			return ( !! get( parentBlockListSettings, [ 'allowedBlocks' ] ) &&
+				get( parentBlockListSettings, [ 'allowedBlocks' ] ).length === 1 );
```
```js
return get( parentBlockListSettings, [ 'allowedBlocks', 'length' ], 0 ) === 1
```

youknowriad commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -102,10 +128,62 @@ class Inserter extends Component {
 
 		return {
 			hasItems: hasInserterItems( rootClientId ),
+			hasOneAllowedItem: hasOneAllowedItem( rootClientId ),
```
If I remember properly, there are multiple reasons that might result in just one item being available in the inserter (in addition to the allowed blocks property). Do we want to change the behavior everytime the inserter only shows one block (rely on the result of getInserterItems) or only apply the behavior once the allowed blocks lenght is 1?

draganescu replied:
yes, not only for the inner blocks but in all cases we should not show a choice of one possible thing, that's my opinion :)

mcsf replied:
What was the verdict here?

draganescu committed:
made a new insertion point selector, some code review refactoring

draganescu committed:
better handling of inserter

draganescu committed:
refactoring and named block insertion

draganescu commented on packages/block-library/src/column/edit.js:
```diff
@@ -63,6 +63,7 @@ function ColumnEdit( {
 				</PanelBody>
 			</InspectorControls>
 			<InnerBlocks
+				allowedBlocks={ [ 'core/paragraph' ] }
```
This is to be removed before merge

draganescu commented:
@mcsf I went a while a ago through the code and your comments. :) I could use a nudge about the current state to get an idea of how far is this from what would be ok :P

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1329,6 +1329,70 @@ export const hasInserterItems = createSelector(
+
+		if ( get( parentBlockListSettings, [ 'allowedBlocks', 'length' ], 0 ) === 1 ) {
+			let name = get( parentBlockListSettings, [ 'allowedBlocks' ] )[ 0 ];
+			name = name.split( '/' )[ 1 ].replace( '-', ' ' );
```
This is effectively attempting to reconstitute a title for the block. We shouldn't do this, for two reasons:

- It's fair to say that the _expectation_ for a function such as `getOneAllowedItemName` would return a proper `blockName`, i.e. `core/paragraph` instead of `paragraph`. Other hunks in this PR also refer to this datum as `blockName`, which is misleading.
- We shouldn't ever attempt to programmatically guess a block's title, neither in English nor in any other locale. Instead, use block types' own declared `title` property.

draganescu replied:
OMG, I was stuck with the `name` thing, that I just didn't realize that I already have the title title to use!

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1329,6 +1329,70 @@ export const hasInserterItems = createSelector(
+	}
+
+	return false;
+};
```
Minor, but prefer the "bail-early" style, which is flatter:

```js
if ( ! rootClientId ) {
  return false;
}

const parentBlockListSettings = …;

if ( ! get( … ) ) {
  return false;
}

const name = …;
```

mcsf reviewed (changes requested):
This is in better shape — thanks for circling back, @draganescu. There are still a couple of pressing things to address.

draganescu committed:
updates to the appender

draganescu committed:
update snapshots

draganescu committed:
update docs

draganescu committed:
default inserter label is used in so many tests

draganescu committed:
fixed allowed blocks test

draganescu committed:
snapshot updated

mcsf commented:
@draganescu, what's the status of this? I see you pushed some changes last weeks, but the tests are failing. Let me know when you'd like a review, or to discuss anything.

draganescu commented:
I would love a new review on this @mcsf I am unsure if the failing tests are related, if they are I'll fix them :)

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,71 @@ export const hasInserterItems = createSelector(
+
+	// If there a selected block, we insert after the selected block.
+	const end = getBlockSelectionEnd( state );
+	if ( ! isAppender && end ) {
```
Logically speaking, we could avoid evaluating `end` if we know upfront that `isAppender` is true. But that's a minor observation, doesn't require action.

mcsf commented on packages/e2e-tests/specs/plugins/inner-blocks-allowed-blocks.test.js:
```diff
@@ -76,6 +77,7 @@ describe( 'Allowed Blocks Setting on InnerBlocks ', () => {
 		) )[ 0 ];
 		await insertButton.click();
 		await insertBlock( 'Image' );
+		await page.click( listSelector );
```
Could you detail this change?

draganescu replied:
As far as I can remember, the focus was not sent to the correct place after the image was inserted which caused the test to fail. I can't remember anything else.

draganescu replied:
Reverted and fixed, see [here](https://github.com/WordPress/gutenberg/pull/16708#issuecomment-543568335)

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -1,29 +1,45 @@
+const defaultRenderToggle = ( { onToggle, disabled, isOpen, blockTitle } ) => {
+	let label = sprintf( _x( 'Add %s', 'directly add the allowed block' ), blockTitle );
+	if ( blockTitle === '' ) {
+		label = 'Add block';
```
Needs i18n.

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -1,29 +1,45 @@
+	let label = sprintf( _x( 'Add %s', 'directly add the allowed block' ), blockTitle );
+	if ( blockTitle === '' ) {
+		label = 'Add block';
+	}
```
I think I'd avoid the pattern of assigning `label` once and then again after looking at `blockTitle`. Instead, I'd define `label` with a `const` and a ternary. Just my suggestion, because it feels awkward to call `_x` in vain and `sprintf` with a potentially missing value.

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +129,47 @@ class Inserter extends Component {
+		};
+	} ),
+	withDispatch( ( dispatch, ownProps, { select } ) => {
+		// eslint-disable-next-line no-restricted-syntax
```
We require each `eslint` disable rule to be justified with a comment. Why disable here?

draganescu replied:
The reason was that there is a rule for no logic in withDispatch but since there is only one thing returned there is no need to extract this so I returned in the object body instead.

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -96,7 +120,7 @@ class Inserter extends Component {
 				expandOnMobile
 				headerTitle={ __( 'Add a block' ) }
-				renderToggle={ this.renderToggle }
+				renderToggle={ toggle }
```
I'd keep `this.renderToggle`, as it's IMO easier to trace variables across the file.

draganescu replied:
yes, good catch, looks like a leftover ...

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +129,47 @@ class Inserter extends Component {
+	} ),
+	withDispatch( ( dispatch, ownProps, { select } ) => {
+		// eslint-disable-next-line no-restricted-syntax
+		function createIfOne() {
```
`createIfOne` isn't very clear for someone with no prior context. How about `autoInsertBlock`? Or something else. :)

draganescu replied:
I think I should rename them all including the selector

draganescu committed:
better naming and removed the need for es-lint disabling

draganescu committed:
improved the inserter label construction

draganescu committed:
improved the doc of getTheOnlyAllowedItem selector

draganescu commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,71 @@ export const hasInserterItems = createSelector(
+ * @param {Object}  state        Editor state.
+ * @param {?string} rootClientId Optional root client ID of block list.
+ *
+ * @return {string} The name of the allowed block.
```
yes, you are right :)

draganescu committed:
reverting test patches becasue patching without understanding is bad, bad, bad - don't do it

draganescu committed:
moved getInsertionIndex out of selectos and back into each component that used it

draganescu committed:
docs generated

draganescu commented:
Okay @mcsf I reverted the test patches ( OMG what was I thinking 😊 ) and found the problem to be in the fact that I moved the getInsertionIndex from the inserter menu component and inserter component and made it a selector in core/block-editor.

Somehow this did not retrieve the proper insertion index although the code was the same and I would expect the state to be the same. 

Nevertheless, adding getInsertionIndex back to the components' withDispatch HOC wrapper worked and now all the tests pass in their original form.

Let me know of further action if needed :)

mcsf commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -478,28 +478,14 @@ export default compose(
-		// it's fine to extract it.
-		// eslint-disable-next-line no-restricted-syntax
-		function getInsertionIndex() {
+		const getInsertionIndex = () => {
```
Thanks for looking (in https://github.com/WordPress/gutenberg/pull/16708#discussion_r335864484) at the rule in question. (For other readers, it's [this one](https://github.com/WordPress/gutenberg/blob/ed24c2f8946c0b0b2fd7783601697b0b5f6502ae/.eslintrc.js#L91-L94).)

In this case, though, I'd rather keep the exception and its comment as is — IMO it's more explicit, whereas turning the function declaration into a `const` feels a little bit like we're tricking ESLint. :)

draganescu replied:
done

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,45 @@ export const hasInserterItems = createSelector(
+	const name = get( parentBlockListSettings, [ 'allowedBlocks', 0 ], 0 );
+
+	return name;
+};
```
A request I should've made earlier, but it's a quick one: could you prefix these selectors with `__unstable` or `__experimental`?

draganescu replied:
and done

draganescu committed:
added experimental labels to new selectors, added es-lint comment back

draganescu committed:
updated docs

draganescu commented:
hi @mcsf I've changed back the function and reverted the comment. Also, the new selectors are all marked experimental now.

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +125,62 @@ class Inserter extends Component {
+					getBlockListSettings,
+				} = select( 'core/block-editor' );
+				const parentBlockListSettings = getBlockListSettings( rootClientId );
+				const isOnlyOneAllowedBlock = get( parentBlockListSettings, [ 'allowedBlocks', 'length' ], 0 ) === 1;
```
Shouldn't this be using the available selector?

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +125,62 @@ class Inserter extends Component {
+					return false;
+				}
+
+				const parentAllowedBlocks = get( parentBlockListSettings, [ 'allowedBlocks' ] );
```
Same as above: should use the selector for the only allowed block.

mcsf commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -478,15 +478,7 @@ export default compose(
-		// This should be an external action provided in the editor settings.
-		const {
-			__experimentalFetchReusableBlocks: fetchReusableBlocks,
-		} = dispatch( 'core/editor' );
```
I don't think we should touch `inserter/menu.js` at all if there is nothing to do here.

mcsf commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +125,62 @@ class Inserter extends Component {
 	withSelect( ( select, { rootClientId } ) => {
-		const { hasInserterItems } = select( 'core/block-editor' );
-
+		const { hasInserterItems, __experimentalHasOnlyOneAllowedInserterItem, __experimentalGetTheOnlyAllowedItem } = select( 'core/block-editor' );
```
Minor, but could split this into one line per var.

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,45 @@ export const hasInserterItems = createSelector(
+};
+
+/**
+ * Gets the name of the only item that may be inserted.
```
```suggestion
 * Returns the name of the only block type that may be inserted, or null if there isn't exactly one allowed type.
```

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,45 @@ export const hasInserterItems = createSelector(
+ *
+ * @return {string?} The name of the allowed block or null.
+ */
+export const __experimentalGetTheOnlyAllowedItem = ( state, rootClientId = null ) => {
```
```suggestion
export const __experimentalGetTheOnlyAllowedBlockType = ( state, rootClientId = null ) => {
```

draganescu replied:
I renamed them both to contain BlockType instead of Item

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,45 @@ export const hasInserterItems = createSelector(
+ * @param {Object}  state        Editor state.
+ * @param {?string} rootClientId Optional root client ID of block list.
+ *
+ * @return {string?} The name of the allowed block or null.
```
```suggestion
 * @return {string?} The name of the allowed block type or null.
```

draganescu committed:
Update packages/block-editor/src/store/selectors.js

Co-Authored-By: Miguel Fonseca <miguelcsf@gmail.com>

draganescu committed:
Update packages/block-editor/src/store/selectors.js

Co-Authored-By: Miguel Fonseca <miguelcsf@gmail.com>

draganescu committed:
refactored and fixed some coding errors

draganescu committed:
small code move

draganescu committed:
small code move

obenland commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -1,29 +1,37 @@
-	/>
-);
+const defaultRenderToggle = ( { onToggle, disabled, isOpen, blockTitle } ) => {
+	const label = blockTitle === '' ? _x( 'Add block', 'Generic label for block inseter button' ) : sprintf( _x( 'Add %s', 'directly add the only allowed block' ), blockTitle );
```
`inseter` typo.

Should we add a translator comment for the placeholder?

obenland commented:
I encountered some oddities when clicking the `+` icon to add more menu items with others already open. Not sure if it's a regression of this PR and/or could be fixed as part of it?

![screen-capture](https://user-images.githubusercontent.com/1398304/67527649-5a190200-f66c-11e9-86b0-95a43ec34b0e.gif)

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +119,67 @@ class Inserter extends Component {
+		const {
+			hasInserterItems,
+			__experimentalHasOnlyOneAllowedBlockType,
+			__experimentalGetTheOnlyAllowedBlockType,
```
I was wondering why two selectors are needed? Could an `__experimentalGetAllowedBlocks` selector serve both purposes?

draganescu replied:
yes, I thought it was better to have a list getter and a boolean selector but I also tried your idea and it is even better :)

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +119,67 @@ class Inserter extends Component {
+	} ),
+	withDispatch( ( dispatch, ownProps, { select } ) => {
+		return {
+			insertTheOnlyAllowedItem: () => {
```
I think 'The' should be dropped from the function name. Probably call it a block instead of an item. Could also potentially use the object shorthand too for minor gains 😄 :
```
insertOnlyAllowedBlock() {
```

talldan commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,45 @@ export const hasInserterItems = createSelector(
+ *
+ * @return {string?} The name of the allowed block type or null.
+ */
+export const __experimentalGetTheOnlyAllowedBlockType = ( state, rootClientId = null ) => {
```
Again, I think 'The' should be dropped from the function name.

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +119,67 @@ class Inserter extends Component {
+				const hasOnlyOneAllowedInserterItem = __experimentalHasOnlyOneAllowedBlockType( rootClientId );
+
+				if ( ! hasOnlyOneAllowedInserterItem ) {
+					return false;
```
Does the return value do anything? If not, I think we can drop it.
```suggestion
					return;
```

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +119,67 @@ class Inserter extends Component {
+					return false;
+				}
+
+				const parentAllowedBlocks = __experimentalGetTheOnlyAllowedBlockType( rootClientId );
```
My understanding is that the function should only return one block? If so, the variable name shouldn't be plural.

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +119,67 @@ class Inserter extends Component {
+				const {
+					insertBlock,
+				} = dispatch( 'core/block-editor' );
+				const insertedBlock = createBlock( parentAllowedBlocks );
```
Major nitpick incoming 😄  - `insertedBlock` is past tense, and I think it should be future as it hasn't been inserted yet. Maybe `blockToInsert`.

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +119,67 @@ class Inserter extends Component {
+
+				const parentAllowedBlocks = __experimentalGetTheOnlyAllowedBlockType( rootClientId );
+
+				function getInsertionIndex() {
```
I see this is also defined in the inserter menu and now used in three places. I wonder if it can be consolidated. I was thinking of whether a new version of `insertBlock` could be made that has this logic built-in.

Maybe one for a separate PR after this is merged?

draganescu replied:
yes, i tried making this a selector but for some reason it didn't work "out of the box", as the position was not the expected one (probably easy to fix if you know what's wrong, but in my case easier to revert).

I would do it in a separate PR and do it like you said above, it's even better :)

talldan commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1284,6 +1284,45 @@ export const hasInserterItems = createSelector(
+export const __experimentalHasOnlyOneAllowedBlockType = ( state, rootClientId = null ) => {
+	if ( rootClientId ) {
+		const parentBlockListSettings = getBlockListSettings( state, rootClientId );
+		return get( parentBlockListSettings, [ 'allowedBlocks', 'length' ], 0 ) === 1;
```
I found it interesting that the `hasInserterItems` selector works quite differently in determining blocks available for insert. I wonder what the reason behind that is.

talldan replied:
Specifically it does this:
```
const hasBlockType = some(
	getBlockTypes(),
	( blockType ) => canIncludeBlockTypeInInserter( state, blockType, rootClientId )
);
```

draganescu replied:
this selector is gone now

talldan reviewed (commented):
Left some comments 😄

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -56,10 +64,17 @@ class Inserter extends Component {
 		} = this.props;
 
-		return renderToggle( { onToggle, isOpen, disabled } );
+		if ( hasOnlyOneAllowedInserterItem ) {
```
Could probably make this simpler and avoid rendering the dropdown completely if there's on a single block. I'm thinking an early return that just renders the `IconButton`.

talldan replied:
Something like:
```
render() {
if ( hasOnlyOneAllowedBlock ) {
  return this.renderToggle( { onToggle: insertTheOnlyAllowedItem } );
}
```

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -1,29 +1,37 @@
+			onClick={ onToggle }
+			className="editor-inserter__toggle block-editor-inserter__toggle"
+			aria-haspopup="true"
+			aria-expanded={ isOpen }
```
These aria properties shouldn't be present when a single block will be directly inserted.

draganescu committed:
removes aria attrs for autoinserted items

draganescu committed:
fixes typo, adds translators comment

draganescu committed:
simplifies the intserter logic

draganescu committed:
fix for the simplification

draganescu committed:
simplifies by using one selector and passing props in compose

draganescu committed:
small code updates

draganescu committed:
lint

draganescu committed:
renamed insertedBlock

draganescu commented:
@mcsf @talldan I have simplified this and we only have one selector now and only one place where we determine if there is just a single allowed block type in the inserter. Let me know about new things, if any :)

draganescu committed:
small doc update

tellthemachines reviewed (commented):
Thanks for all your hard work on this! Just one comment/question below, and another question 😬 : do you plan to address [this comment](https://github.com/WordPress/gutenberg/pull/16708#issuecomment-546295462) as part of this PR or as a separate one? (Might make more sense as a separate one.)

draganescu commented:
@tellthemachines I think #18104 should be implemented in a separate PR because it is also touching the block menu.

draganescu commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -105,10 +122,72 @@ class Inserter extends Component {
+
+		const allowedBlocks = __experimentalGetAllowedBlocks( rootClientId );
+
+		const hasOnlyOneAllowedInserterBlockType = allowedBlocks && ( get( allowedBlocks, [ 'length' ], 0 ) === 1 );
```
this is a bug introduced by all the refactoring, will check!

draganescu replied:
It wasn't a but in my code, but the default button appender was simply printing __('Add blocl') instead of the label I have supplied. I have also added in this PR a tooltip for the default button appender.

talldan reviewed (commented):
Feels pretty close!

draganescu committed:
adds tooltip to the default button appender

draganescu committed:
refactores for more self documenting varnames

draganescu commented:
@talldan now it's even better! :) In the meantime following @tellthemachines ' observation I have also added a tooltip to the default button appender which I think is nice.

talldan commented on packages/block-editor/src/components/button-block-appender/index.js:
```diff
@@ -21,17 +21,31 @@ function ButtonBlockAppender( { rootClientId, className } ) {
+								<Icon icon="insert" />
+							</Button>
+						</Tooltip>
+					);
```
Hmm, this button with an icon and a tooltip is now looking exactly like an `IconButton` 😄 
https://github.com/WordPress/gutenberg/blob/master/packages/components/src/icon-button/index.js

talldan reviewed (approved):
I think this is working well now, and the code is a lot tidier.

@draganescu It'd be good to see some small follow-up PRs that:
- Add some tests for the new selector
- Try refactoring `ButtonBlockAppender` to use an `IconButton`.
- Try deduplicating the `getInsertionIndex` for the various `insertBlock` usages.

talldan commented:
One other bit of feedback is that I tried disabling all blocks in the block manager apart from the paragraph. I was expecting all the Inserters in the editor to then only insert the paragraph directly without showing the menu, but it didn't work. It's a bit of a weird one, so I don't consider it a blocker, but I wonder if the selector logic needs some work so that this truly works dynamically.
